database/schedule/list.go

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2023 Target Brands, Inc. All rights reserved.
+//
+// Use of this source code is governed by the LICENSE file in this repository.
+
+package schedule
+
+import (
+	"github.com/go-vela/types/constants"
+	"github.com/go-vela/types/database"
+	"github.com/go-vela/types/library"
+)
+
+// ListSchedules gets a list of all schedules from the database.
+func (e *engine) ListSchedules() ([]*library.Schedule, error) {
+	e.logger.Trace("listing all schedules from the database")
+
+	// variables to store query results and return value
+	count := int64(0)
+	s := new([]database.Schedule)
+	schedules := []*library.Schedule{}
+
+	// count the results
+	count, err := e.CountSchedules()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I'm curious if this is necessary. It doesn't look like we use it other than short circuiting an empty table, and the `COUNT` query is fairly taxing (at least in Postgres): https://wiki.postgresql.org/wiki/Introduction_to_VACUUM,_ANALYZE,_EXPLAIN,_and_COUNT#COUNT.28.2A.29

Reply:
To provide context, this was adopted from an existing pattern we already use for `builds`:

https://github.com/go-vela/server/blob/3aa414e46ced00ce3019f0d58120801a7c4cd233/database/postgres/build_list.go#L89-L113

 And this code follows the same pattern used in the other database functions to list resources e.g.

* https://github.com/go-vela/server/blob/main/database/hook/list.go#L14
* https://github.com/go-vela/server/blob/main/database/repo/list.go#L14
* https://github.com/go-vela/server/blob/main/database/user/list.go#L14

In general, this need for the count stems from usage for API pagination.

Before I started the DB refactor efforts, we still called the count for API but did so in a separate function i.e.

https://github.com/go-vela/server/blob/8c2fdb27d4a8ab8eb55e272a9135c6939c1ec28b/api/service.go#L241-L259

I decided it was worth adopting the pattern setup for `builds` since it reduces the need to call them separately.

Reply:
Also, I thought it was worth highlighting a statement from the document you linked:

> If you are using count(*), the database is free to use any column to count, which means it can pick the smallest covering index to scan (note that this is why count(*) is much better than count(some_field), as long as you don't care if null values of some_field are counted). Since indexes often fit entirely in memory, this means count(*) is often very fast.

In general, calling `count(*)` for a specific table isn't very taxing. Could you elaborate on your concern?

Are you facing issues with this for other tables and that's why you're mentioning it here?

Reply:
To be clear, I don't really have any _issues_ with the implementation, nor are we experiencing performance problems. Though, `COUNT` queries across a specific table without any `WHERE` clause related to an index are relatively slow (or slower than you'd expect is probably a better descriptor): https://wiki.postgresql.org/wiki/Slow_Counting. 

I'm fine moving this conversation elsewhere as a potential reconsideration of design since it's endemic across the code base, as you pointed out 👍 

Just trying to give all my thoughts so I don't forget 😃

Reply:
sure 👍 perhaps it's worth opening an issue for this so it  doesn't get forgotten?

Reply:
https://github.com/go-vela/community/issues/810